File: app/src/main/java/igrek/forceawaken/service/ringtone/RingtoneManagerService.java
package igrek.forceawaken.service.ringtone;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import javax.inject.Inject;

import igrek.forceawaken.dagger.DaggerIOC;
import igrek.forceawaken.domain.ringtone.Ringtone;
import igrek.forceawaken.logger.Logger;
import igrek.forceawaken.logger.LoggerFactory;
import igrek.forceawaken.service.filesystem.ExternalCardService;
import igrek.forceawaken.service.filesystem.FilesystemService;

public class RingtoneManagerService {
	
	@Inject
	FilesystemService filesystemService;
	
	private Logger logger = LoggerFactory.getLogger();
	private ExternalCardService externalCardService;
	
	private Random random = new Random();
	
	public RingtoneManagerService(ExternalCardService externalCardService) {
		this.externalCardService = externalCardService;
		DaggerIOC.getFactoryComponent().inject(this);
		filesystemService.ensureAppDataDirExists();
	}
	
	public Ringtone getRandomRingtone() {
		List<Ringtone> ringtones = getAllRingtones();
		return ringtones.get(random.nextInt(ringtones.size()));
	}
	
	public List<Ringtone> getAllRingtones() {
		String ringtonesPath = getExternalStorageDirectory() + "/Android/data/igrek.forceawaken/ringtones";
		File ringtonesDir = new File(ringtonesPath);
		if (!ringtonesDir.exists()) {
			logger.warn("ringtones dir does not exist");
			ringtonesDir.mkdirs();
		}
		List<Ringtone> ringtones = new ArrayList<>();
		for (File file : ringtonesDir.listFiles()) {
			String name = getRingtoneName(file);
			ringtones.add(new Ringtone(file, name));
		}
		return ringtones;
	}
	
	private String getExternalStorageDirectory() {
		return externalCardService.getExternalSDPath();
	}
	
	private String getRingtoneName(File ringtone) {
		return ringtone.getName().replaceAll("\\.mp3$", "");
	}
	
}

File: app/src/main/java/igrek/forceawaken/ringtone/RingtoneManagerService.kt
package igrek.forceawaken.ringtone

import android.app.Activity
import igrek.forceawaken.info.logger.Logger
import igrek.forceawaken.info.logger.LoggerFactory
import igrek.forceawaken.inject.LazyExtractor
import igrek.forceawaken.inject.LazyInject
import igrek.forceawaken.inject.appFactory
import igrek.forceawaken.system.filesystem.ExternalCardService
import java.io.File
import java.util.*

class RingtoneManagerService(
        externalCardService: LazyInject<ExternalCardService> = appFactory.externalCardService,
        activity: LazyInject<Activity> = appFactory.activity,
) {
    private val externalCardService by LazyExtractor(externalCardService)
    private val activity by LazyExtractor(activity)

    private val logger: Logger = LoggerFactory.logger

    private val random = Random()

    val randomRingtone: Ringtone
        get() {
            val ringtones = allRingtones
            return ringtones[random.nextInt(ringtones.size)]
        }

    val allRingtones: List<Ringtone>
        get() {
            val appDataDir = activity.getExternalFilesDir("data")?.absolutePath
            val ringtonesPath = "$appDataDir/ringtones"
            val ringtonesDir = File(ringtonesPath)
            if (!ringtonesDir.exists()) {
                logger.warn("ringtones dir does not exist: $ringtonesPath")
                ringtonesDir.mkdirs()
            }
            val ringtones: MutableList<Ringtone> = ArrayList()
            val files = ringtonesDir.listFiles()
                    ?: throw RuntimeException("Listing files got null: $ringtonesPath")
            files.filter { it.isFile }.forEach { file ->
                val name = getRingtoneName(file)
                ringtones.add(Ringtone(file, name))
            }
            return ringtones
        }

    private val externalStorageDirectory: String?
        get() = externalCardService.externalSDPath

    private fun getRingtoneName(ringtone: File): String {
        return ringtone.name.replace("\\.mp3$".toRegex(), "")
    }


}